### ISSUE TITLE
Validation rules/functions to miden assembly

### ISSUE THREAD

**Issue by calummoore** [polybase/polylang#18]
Compile our validation rules/functions language into miden assembly.

**Comment by mateuszmlc** [comment]
A couple issues:

## You can only access the top 16 stack values

If you say have 9 u64 variables, you can't access the 9th (2 stack values per 1 u64).

Possible solution: Statically allocate (at compile time) an address for each variable, and use mem_load, mem_store to load or change it. These are both only 1 cycle, so it would be pretty cheap and convenient.

<details>
<summary>Example solution</summary>

```
# let i = 0; # address for i: 0x1
# let y = i + 1; # address for y: 0x2

push.0
mem_store.0x1
drop

mem_load.0x1
push.1
add
mem_store.0x2
drop
```

Optimizations could be made to for example replace `mem_store.0x1, drop, mem_load.0x1` with just `mem_store.0x1`.

</details>

## Miden has no break/jump/return instruction

This makes it hard to implement e.g. a return in the middle of a function, throw, or break in the middle of a loop.

We could limit return to only appear as the last statement in a function.
Break is not crucial and could be emulated by the user with a variable + if statements.

Throw is the most expensive, every call to a throwable function will need an if check.

<details>

<summary>Example solution for break in loops</summary>

```
# let i = 0;
# while (true) {
#   if (i > 0) {
#     break;
#   }
#   
#   i = i + 1;
# }

begin
    push.0 # value for variable i
    mem_store.1

    push.1
    mem_store.2 # no_break

    while.true
        mem_load.1
        push.0
        gt
        if.true
            push.0
            mem_store.2 # no_break = 0, break
            drop
        end

        mem_load.2
        if.true
            mem_load.1
            push.1
            add
            mem_store.1
            drop
            push.1
        end

        mem_load.2
    end
end
```

</details>

## Strings are expensive

Storing/Loading strings is expensive. Storing just one character takes 6 cycles (although this could be optimized to 7 characters per 6 cycles).

Possible solution: Only deal with length (for min/max) + hashes of strings (for equality check). This would make it impossible to implement the `regex` field validation function.

<details>
<summary>Example code for loading a string</summary>

```
    # string 1 = "abc"
    push.3 # string 1 length
    push.0 # string 1 start address

    push.97 # 'a'
    push.0 # offset
    dup.2 # copy start address
    add # add start address + offset
    mem_store # store 'a' at start address + offset
    drop

    push.98 # 'b'
    push.1 # offset
    dup.2 # copy start address
    add # add start address + offset
    mem_store # store 'b' at start address + offset
    drop

    push.99 # 'c'
    push.2 # offset
    dup.2 # copy start address
    add # add start address + offset
    mem_store # store 'c' at start address + offset
    drop
```

</details>

## Recursive functions are not possible

A procedure can't execute itself (or any function from it's stack trace).
This will make it difficult to implement a correct `regex` function, but not impossible with some restrictions on the regex (like only allowing a single `*` wildcard).

## How to implement a `throw`?

Since we want to show the user a friendly error message, we can't use assert.

Can we succeed with generating a proof on a `throw` and return an error value on the stack?
Say something like:

```
begin
  push.1 # is_error = true
  push.29 # "You don't have enough balance".length
  push.89 # 'Y'
  push.111 # 'o'
  ... 
end
```

## The VM is slow

This limits possible solutions for all the above issues. It limits the shortcuts we can take, because we don't want proofs to take too long to generate.
It's over 4x slower on WASM. Optimizations are possible but limited, due to WASM's small instruction set (only 128bit SIMD, it doesn't help much).

**Comment by mateuszmlc** [comment]
A general shortcut we can take:
We could for now skip implementing compilation of user defined functions and just generate the default set function, with hand-written miden assembly for things like field validation functions.
Once the compiler is finished, we would generate this code instead of writing it by hand (although I think performance critical functions should be in miden assembly).

**Comment by calummoore** [comment]
> You can only access the top 16 stack values

Your suggestions seems reasonable for now. We could look at optimising this further by digging into the miden core.

> Miden has no break/jump/return instruction

I wonder if this is because they verify the circuit as part of the proof, so if the circuit is not complete it would mess with that verification. In that case, we might be able to add an instruction that basically fills everything up with 0s. I can check with my friend to see if she has some insight on this.

> Strings are expensive

I feel like there must be a better way to handle strings, maybe using advice tape? Let me have a think about this.

> Recursive functions are not possible

For now, let's disable this functionality. If we need to temporarily drop regex requirement, then let's do that.

One way to handle recursive functions would be to break them out into separate proof executions. And then have miden verify the proof. We'd basically need to create our own [chiplet](https://maticnetwork.github.io/miden/design/chiplets/main.html) for that, but we could then ensure it is highly optimised.

> How to implement a throw?

If something is going to throw, then we don't actually need to prove that it will throw as the proof would never be sent anywhere. So in that case, we should use assert/fail to make sure they don't get a proof.

We then need to figure out a way to provide a good stack/error message. I think we could do that by modifying the miden compiler/generator to extract the stack at the point of failure, so we can construct the throw message (i.e. whatever variable is provided `error()`). It would also be good if we could provide a stack trace using the output from our parser and some kind of annotation meta.

Another option would be to implement a non-snark execution engine that runs through the code first, and ensures that it is valid, before using the more expensive snark approach. This could result in quite a bit of duplicated work though!


> The VM is slow

* Threads should be possible within the browser - https://webassembly.org/roadmap/
* Creating custom chiplets - as I mentioned above, miden supports custom chiplets, so we could add our own to get even greater performance for specific use cases
* If we go down the option of having proofs compiled on a server (for perf), I would like it to be optional (so you can choose to have someone create the proof for you, or you can create it yourself). The server side optimisations could then be part of the same code base, with different compilation flag that includes optimisations.


> A general shortcut we can take

For now, let's try to keep working on the compiler, as I think we are actually making really good progress and discovering really important trade offs. The faster we learn about these trade offs, the better our future decision making can be.

**Comment by mateuszmlc** [comment]
> I wonder if this is because they verify the circuit as part of the proof, so if the circuit is not complete it would mess with that verification. In that case, we might be able to add an instruction that basically fills everything up with 0s. I can check with my friend to see if she has some insight on this.

For now I can implement it as I did in the loop code example, where I have a boolean, flip it on break/return and guard statements after break/return with `if`, which won't execute if the boolean was flipped.

> We then need to figure out a way to provide a good stack/error message. I think we could do that by modifying the miden compiler/generator to extract the stack at the point of failure, so we can construct the throw message (i.e. whatever variable is provided error()). It would also be good if we could provide a stack trace using the output from our parser and some kind of annotation meta. 

We might actually be able to use `miden_processor::execute_iter` for this:

> The `execute_iter()` function returns a `VmStateIterator` which can be used to iterate over the cycles of the executed program for debug purposes. In fact, when we execute a program using this function, a lot of the debug information is retained and we can get a precise picture of the VM's state at any cycle. Moreover, if the execution results in an error, the `VmStateIterator` can still be used to inspect VM states right up to the cycle at which the error occurred.

So we could push the error to stack before asserting.
This might be slow in practice, I'll have to check that.

> I feel like there must be a better way to handle strings, maybe using advice tape? Let me have a think about this.

I think no matter how smart we do it, at best we can read 7 bytes (stored in one field element) per cycle. For short strings this is fine, but long strings (like say base64 encoded profile picture) it would quickly become expensive.

**Comment by calummoore** [comment]
Closing in favour of https://github.com/polybase/polylang/issues/27
